File: drivers.cpp
#include <iostream>
#include <wiringPi.h>
#include <softPwm.h>
#include <thread>
using namespace std;

// #define LED 0 matches with ASUS_GPIO 164! This can be checked with command 'sudo gpio readall'.
struct Position {
	float phi, theta;
};
class Motors {
public:
	const int MotorAPWM = 0;
	const int MotorA0 = 2;
	const int MotorA1 = 3;

	const int MotorBPWM = 12;
	const int MotorB0 = 13;
	const int MotorB1 = 14;

	std::thread execution;
	bool shutdown = false;

	Position target;
	Position current;
	bool moveA = false;
	bool moveB = false;

	const float epsilon = 1e-5;

	Motors(){
		wiringPiSetup ();
		pinMode (MotorAPWM, OUTPUT);
		softPwmCreate(MotorAPWM, 0, 100);
		pinMode (MotorA0, OUTPUT);
		pinMode (MotorA1, OUTPUT);

		pinMode (MotorBPWM, OUTPUT);
		softPwmCreate(MotorBPWM, 0, 100);
		pinMode (MotorB0, OUTPUT);
		pinMode (MotorB1, OUTPUT);
		execution = std::thread(&Motors::loop,this);
	}
	~Motors(){
		shutdown = true;
		execution.join();
	}
	void stop(){
		target = current;
		moveA = moveB = false;
		softPwmWrite(MotorAPWM, 100); 
		softPwmWrite(MotorBPWM, 100); 
		digitalWrite (MotorA0, HIGH);
		digitalWrite (MotorA1, HIGH);
		digitalWrite (MotorB0, HIGH);
		digitalWrite (MotorB1, HIGH);
	}
	void sendTarget(Position pos){
		softPwmWrite(MotorAPWM, 100); 
		softPwmWrite(MotorBPWM, 100); 
		target = pos;
		moveA = moveB = true;

	}
private:
	void loop(){
		while(!shutdown){
			if(moveA or moveB){
				float dPhi = target.phi - current.phi;
				float dTheta = target.theta - current.theta;
				if(dPhi < epsilon){
					digitalWrite (MotorA0, LOW);
					digitalWrite (MotorA1, LOW);
					moveA = false;
				}
				if(dTheta < epsilon){
					digitalWrite (MotorB0, LOW);
					digitalWrite (MotorB1, LOW);
					moveB = false;
				}
				if(moveA){
					if(dPhi > 0){
						digitalWrite (MotorA0, HIGH);
						digitalWrite (MotorA1, LOW);
					}
					else {
						digitalWrite (MotorA0, LOW);
						digitalWrite (MotorA1, HIGH);
					}
					softPwmWrite(MotorAPWM, getSpeed(dPhi)); 
				}
				if(moveB){
					if(dTheta > 0){
						digitalWrite (MotorB0, HIGH);
						digitalWrite (MotorB1, LOW);
					}
					else {
						digitalWrite (MotorB0, LOW);
						digitalWrite (MotorB1, HIGH);
					}
					softPwmWrite(MotorBPWM, getSpeed(dTheta)); 
				}
				delay(20);
			}
			else {
				delay(250);
			} 
		}
	}
	int getSpeed(float dx){
		return 75;
	}

};

int main (void)
{
	Motors motors;
	for (;;)
	{
		digitalWrite (motors.MotorA0, HIGH);
		digitalWrite (motors.MotorA1, LOW);
				digitalWrite (motors.MotorB0, LOW);
		digitalWrite (motors.MotorB1, LOW);
		cout << "tik"<< endl;

		for(int i  = 0; i <= 200; i++){
			softPwmWrite(motors.MotorAPWM, abs(100 - i)); 
			delay (1);

		}
		digitalWrite (motors.MotorA0, LOW);
		digitalWrite (motors.MotorA1, HIGH);
		digitalWrite (motors.MotorB0, LOW);
		digitalWrite (motors.MotorB1, LOW);
		cout << "tok"<< endl;
		for(int i  = 0; i <= 200; i++){
			softPwmWrite(motors.MotorAPWM, abs(100 - i)); 
			delay (1);
		}
				digitalWrite (motors.MotorA0, LOW);
		digitalWrite (motors.MotorA1, LOW);
		digitalWrite (motors.MotorB0, HIGH);
		digitalWrite (motors.MotorB1, LOW);
			for(int i  = 0; i <= 200; i++){
			softPwmWrite(motors.MotorAPWM, abs(100 - i)); 
			delay (1);
		}
				digitalWrite (motors.MotorA0, LOW);
		digitalWrite (motors.MotorA1, LOW);
		digitalWrite (motors.MotorB0, LOW);
		digitalWrite (motors.MotorB1, HIGH);
			for(int i  = 0; i <= 200; i++){
			softPwmWrite(motors.MotorAPWM, abs(100 - i)); 
			delay (1);
		}
				digitalWrite (motors.MotorA0, LOW);
		digitalWrite (motors.MotorA1, LOW);
		digitalWrite (motors.MotorB0, LOW);
		digitalWrite (motors.MotorB1, LOW);
			for(int i  = 0; i <= 400; i++){
			softPwmWrite(motors.MotorAPWM, abs(100 - i)); 
			delay (1);
		}
	}
    return 0;
}

File: drivers/drivers.py
import RPi.GPIO as GPIO
import threading
from time import sleep

motor_a_pwm = 12
motor_a0 = 13
motor_a1 = 15

motor_b_pwm = 16
motor_b0 = 18
motor_b1 = 19

frequency = 1

GPIO.setmode(GPIO.BOARD)

GPIO.setup(motor_a_pwm, GPIO.OUT)

pwm_a = GPIO.PWM(motor_a_pwm, frequency)

GPIO.setup(motor_a0, GPIO.OUT)
GPIO.setup(motor_a1, GPIO.OUT)

GPIO.setup(motor_b_pwm, GPIO.OUT)

pwm_b = GPIO.PWM(motor_b_pwm, frequency)

GPIO.setup(motor_b0, GPIO.OUT)
GPIO.setup(motor_b1, GPIO.OUT)

shutdown = False

epsilon = 0.01

target = (0.0, 0.0)
current = (0.0, 0.0)


def get_speed(dx):
    return 75.0  # to calculate


def sendTarget(pos):
    (phi, theta) = pos
    if len(pos) == 2:
        target = pos
        return 0
    return 1


def loop():
    while not shutdown:
        (target_phi, target_theta) = target
        (current_phi, current_theta) = current
        
        dPhi = target_phi - current_phi
        dTheta = target_theta - current_theta

        moveA = False
        moveB = False
        
        if dPhi < epsilon:
            GPIO.output(motor_a0, GPIO.LOW)
            GPIO.output(motor_a1, GPIO.LOW)
        if dTheta < epsilon:
            GPIO.output(motor_b0, GPIO.LOW)
            GPIO.output(motor_b1, GPIO.LOW)
        if moveA and dPhi > 0:
            GPIO.output(motor_a0, GPIO.HIGH)
            GPIO.output(motor_a1, GPIO.LOW)
            pwm_a.ChangeDutyCycle(get_speed(dPhi))
        elif moveA:
            GPIO.output(motor_a0, GPIO.LOW)
            GPIO.output(motor_a1, GPIO.HIGH)
            pwm_a.ChangeDutyCycle(get_speed(dPhi))
        if moveB and dTheta > 0:
            GPIO.output(motor_b0, GPIO.HIGH)
            GPIO.output(motor_b1, GPIO.LOW)
            pwm_b.ChangeDutyCycle(get_speed(dTheta))
        elif moveA:
            GPIO.output(motor_b0, GPIO.LOW)
            GPIO.output(motor_b1, GPIO.HIGH)
            pwm_b.ChangeDutyCycle(get_speed(dTheta))
        
        sleep(0.05)
    
    pwm_a.ChangeDutyCycle(100)
    pwm_b.ChangeDutyCycle(100)

    GPIO.output(motor_a0, GPIO.HIGH)
    GPIO.output(motor_a1, GPIO.HIGH)
    GPIO.output(motor_b0, GPIO.HIGH)
    GPIO.output(motor_b1, GPIO.HIGH)

    pwm_a.stop()
    pwm_b.stop()
    GPIO.cleanup()




def stop(t):
    global shutdown
    shutdown = True
    t.join()

def start():
    t = threading.Thread(target=loop)
    t.start()
    return t



t = start()
sleep(1)
stop(t)